src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -152,6 +152,14 @@ addNewFilterStepOrThrow(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes,
     /// New filter column is the first one.
     String split_filter_column_name = split_filter.getOutputs().front()->result_name;
 
+    // If no new columns added, filter just used one of the input columns as-is and moved it to the front, move it back to keep aggregation key in order.
+    if (const auto & input = node.children.at(0)->step->getOutputHeader(); split_filter.getOutputs().size() == input.columns())
+    {
+        auto pos = input.getPositionByName(split_filter_column_name);
+        if (pos != 0)
+            std::rotate(split_filter.getOutputs().begin(), split_filter.getOutputs().begin() + 1, split_filter.getOutputs().begin() + pos + 1);
+    }
+
```

Comment:
Maybe let's add an explicit check that all columns at the end of the day go in the expected order. Who knows if this bug can also be reproduced with some other query

Reply:
We have this check, but later and only in debug mode
https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Aggregator.cpp#L342